### PR TITLE
Add TON Address

### PR DESCRIPTION
**Pull request by hewigovens**
Clean and format code from: https://github.com/trustwallet/wallet-core/pull/616, also rename from Telegram to TON

What's next

- [x] initiate address with public key and master/base chain id 
- [x] Return Bounceable address format by default
- [ ] Expose method to return user friendly (base64) Non-bounceable / raw address
- [x] Cell builder to build StateInit (and external message) from public key
- [x] move base64 url encoding to class `Base64`

### CODE REVIEW

**Comment by catenocrypt** [comment]
Resolved conflict; rebased to master, include splitting of CoinType tests.

**Comment by catenocrypt** [comment]
Finalized, ready for review/merge.  
Note: Address part is complete, signing part is partial, only account init msg is supported.

File: include/TrustWalletCore/TWTONAddress.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+TW_EXPORT_STATIC_METHOD
+bool TWTONAddressIsValidString(TWString *_Nonnull string);
+
+/// Creates an address from a string representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWTONAddress *_Nullable TWTONAddressCreateWithString(TWString *_Nonnull string);
+
+/// Creates an address from a public key.
+TW_EXPORT_STATIC_METHOD
+struct TWTONAddress *_Nonnull TWTONAddressCreateWithPublicKey(struct TWPublicKey *_Nonnull publicKey);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWTONAddress.h]: can we let caller to choose chain id? `TWTONAddressCreateWithPublicKey(publicKey, chainId);

File: include/TrustWalletCore/TWTONSigner.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+TW_EXTERN_C_BEGIN
+
+/// Helper class to sign TON transactions.
+TW_EXPORT_CLASS
+struct TWTONSigner;
+
+/// Signs a transaction.
+//TW_EXPORT_STATIC_METHOD
+//TW_TON_Proto_SigningOutput TWTONSignerSign(TW_TON_Proto_SigningInput input);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWTONSigner.h]: let's enable this, account init msg is also kind of signing input, we can define different messages in .proto

File: src/TON/Address.cpp [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+Address::Address(const std::string& address) {
+    bool valid = parseAddress(address, *this);
+
+    // Ensure address is valid
+    if (!valid) {
+        throw std::invalid_argument("Invalid address data");
+    }
+}
+
+Address::Address(const PublicKey& publicKey) {
```
Comment (hewigovens) [inline src/TON/Address.cpp]:
```suggestion
Address::Address(const PublicKey& publicKey, WorkchainId_t chainId) {
```

File: src/TON/Address.cpp [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    auto hash = stateInit.hash();
+
+    // fill members
+    workchainId = Workchain::defaultChain();
+    addrBytes = hash;
+    isBounceable = true;
+    isTestOnly = false;
+}
+
+bool Address::isValid(const std::string& address) {
```
Comment (hewigovens) [inline src/TON/Address.cpp]:
```suggestion
bool Address::isValid(const std::string& address, WorkchainId_t chainId) {
```

File: src/TON/Address.cpp [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    addr_inout.addrBytes = parse_hex(addressStr);
+
+    addr_inout.isBounceable = true;
+    addr_inout.isTestOnly = false;
+
+    return true;
+}
+
+bool Address::parseUserAddress(const std::string& addressStr_in, Address& addr_inout) {
```
Comment (hewigovens) [inline src/TON/Address.cpp]: maybe `parseBase64Address`?

File: src/TON/Address.cpp [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    Address addr;
+    bool isValid = parseAddress(address, addr);
+    if (addr.workchainId != Workchain::defaultChain()) {
+        // not the right chain, refuse
+        return false;
+    }
+    return isValid;
+}
+
+bool Address::parseAddress(const std::string& addressStr_in, Address& addr_inout) {
```
Comment (hewigovens) [inline src/TON/Address.cpp]:
```suggestion
bool Address::parseAddress(const std::string& address, Address& outAddr) {
```

File: src/TON/Address.cpp [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    Cell stateInit = Contract::createStateInit(publicKey);
+
+    // compute hash of stateInit
+    auto hash = stateInit.hash();
+
+    // fill members
+    workchainId = Workchain::defaultChain();
+    addrBytes = hash;
+    isBounceable = true;
+    isTestOnly = false;
```
Comment (hewigovens) [inline src/TON/Address.cpp]: do we still need this flag?

File: src/TON/Address.cpp [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    default:
+        return false; // invalid chain
+    }
+
+    // 32 bytes address
+    addr_inout.addrBytes = Data(AddressLength);
+    std::copy(bytes.begin() + 2, bytes.begin() + 2 + AddressLength, addr_inout.addrBytes.begin());
+
+    // check CRC
+    uint16_t crcGiven = static_cast<uint16_t>(bytes[34] << 8) + bytes[35];
```
Comment (hewigovens) [inline src/TON/Address.cpp]: we can extract a computeChecksum method

File: src/TON/Cell.cpp [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    // child cells
+    for(auto c: _cells) {
+        c->serializeOwn(data_inout, false);
+    }
+
+    if (mode & SerializationMode::WithCRC32C) {
+        // CRC32-C
+        using crc_32c_type = boost::crc_optimal<32, 0x1EDC6F41, 0xFFFFFFFF, 0xFFFFFFFF, true, true>;
```
Comment (hewigovens) [inline src/TON/Cell.cpp]: can we move this magic to `Crc.h/.cpp`?

File: src/TON/Cell.h [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+protected:
+    void appendBitsAligned(const TW::Data& data_in, size_t sizeBits);
+    void appendBitsNotAligned(const TW::Data& data_in, size_t sizeBits);
+
+private:
+    Data _data;
+    size_t _sizeBits;
+};
+
+/**
```
Comment (hewigovens) [inline src/TON/Cell.h]: let's move these comments to `README.md` in src/TON ?

File: include/TrustWalletCore/TWTONAddress.h (hunk 1 above)
Comment (catenocrypt) [reply include/TrustWalletCore/TWTONAddress.h]: I'm not sure if this is needed to be exposed to the outside completely.  For now I exposed on the C++ class only.

File: src/TON/Address.cpp (hunk 3 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: OK

File: src/TON/Address.cpp (hunk 4 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: OK

File: src/TON/Address.cpp (hunk 5 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: The TON terminology is User and Raw format.

File: src/TON/Address.cpp (hunk 6 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: done, minor

File: src/TON/Address.cpp (hunk 7 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: I'm not sure, I leave it

File: src/TON/Address.cpp (hunk 8 above)
Comment (catenocrypt) [reply src/TON/Address.cpp]: One place, one line usage, I leave here.

File: src/TON/Cell.h (hunk 10 above)
Comment (catenocrypt) [reply src/TON/Cell.h]: Good idea, done.

File: src/TON/Cell.cpp (hunk 9 above)
Comment (catenocrypt) [reply src/TON/Cell.cpp]: Extracted into separate method.  Used only from one place so far.

**Pull request merged**